### ISSUE TITLE
Logical replication slot is dropped which leads to the one's inoperability

### ISSUE THREAD

**Issue by akalgush** [patroni/patroni#3237]
### What happened?

Logical replication subscription create a permanent slot nowadays (since version 14) for sybcronization purposes. I met a scenario where logical replication was not functioning due to a deletion of a permanent inactive slot. To reproduce this there must be another slot - active and unknown to patroni (not being mentioned in section 'slots'). Turn patroni to pause mode then resume it back. Patroni starts to try to delete the unknown slot on constant basis. If any permanent inactive slot (like the ones for initial logical syncronization) is created at this moment, patroni just delete it.

It's not clear from docs, release notes and code what is the intended behaviour for patroni to manage slots. From one side any unknown slot (but temporary one) is about to be dropped, from another side it is often not dropped until restart. It is not clear at which moment a slot is expected to be dropped which is confusing. It would be great to clarify this.

### How can we reproduce it (as minimally and precisely as possible)?

-- create subscription on receiver side
CREATE SUBSCRIPTION test_sub1
CONNECTION 'host=source_host port=source_port user=postgres dbname=postgres sslmode=verify-ca sslcert=/pg_ssl/client.crt sslkey=/pg_ssl/client.key sslrootcert=/pg_ssl/root.crt'
PUBLICATION test_pub
WITH (copy_data = true, create_slot=true, enabled=true, slot_name=test_sub1);

postgres=# select * from pg_replication_slots;
 slot_name |  plugin  | slot_type | datoid | database | temporary | active | active_pid | xmin | catalog_xmin | restart_lsn | confirmed_flush_lsn | wal_status | safe_wal_size 
-----------+----------+-----------+--------+----------+-----------+--------+------------+------+--------------+-------------+---------------------+------------+---------------
 test_sub1 | pgoutput | logical   |  14422 | postgres | f         | t      |    2249639 |      |   2301612206 | 19/B2002108 | 19/B20021B8         | reserved   |              
(1 row)

$ patronictl -c /etc/patroni/postgres.yml pause
Success: cluster management is paused

-- restart is not needed, doing this to double-check that patroni does not delete it in pause-mode
$ pg_ctl restart
waiting for server to shut down.... done
server stopped
waiting for server to start....2024-12-04 10:45:08 MSK [2249602]: [1-1] app=,user=,db=,client=,type=postmaster LOG:  Default auto-detected value for pg_stat_kcache.linux_hz is 1000000
2024-12-04 10:45:08 MSK [2249602]: [2-1] app=,user=,db=,client=,type=postmaster LOG:  pg_ouline: outlines disabled
2024-12-04 10:45:08 MSK [2249602]: [3-1] app=,user=,db=,client=,type=postmaster LOG:  redirecting log output to logging collector process
2024-12-04 10:45:08 MSK [2249602]: [4-1] app=,user=,db=,client=,type=postmaster HINT:  Future log output will appear in directory "postgresql_log_dir".
 done
server started

$ patronictl -c /etc/patroni/postgres.yml resume
Success: cluster management is resumed

-- then patroni tries to delete the unknown slot test_sub1 every 10 sec (loop_wait)
Dec  4 10:46:12 source_host patroni.bin[1309802]: 2024-12-04 10:46:12,797 ERROR: Failed to drop replication slot 'test_sub1'
Dec  4 10:46:12 source_host patroni.bin[1309802]: 2024-12-04 10:46:12,861 INFO: no action. I am (source_host) the leader with the lock
Dec  4 10:46:22 source_host patroni.bin[1309802]: 2024-12-04 10:46:22,689 INFO: Lock owner: source_host; I am source_host
Dec  4 10:46:22 source_host patroni.bin[1309802]: 2024-12-04 10:46:22,757 ERROR: Failed to drop replication slot 'test_sub1'

-- create slot with the name like logical syncronization slot
postgres=# SELECT * FROM pg_create_logical_replication_slot('pg_19732_sync_17854_7442686389907653644', 'pgoutput', 'false');
                slot_name                |     lsn     
-----------------------------------------+-------------
 pg_19732_sync_17854_7442686389907653644 | 19/B2002228
(1 row)

-- then the logical syncronization slot was immediatelly dropped
postgres=# select * from pg_replication_slots;
 slot_name |  plugin  | slot_type | datoid | database | temporary | active | active_pid | xmin | catalog_xmin | restart_lsn | confirmed_flush_lsn | wal_status | safe_wal_size 
-----------+----------+-----------+--------+----------+-----------+--------+------------+------+--------------+-------------+---------------------+------------+---------------
 test_sub1 | pgoutput | logical   |  14422 | postgres | f         | t      |    2249639 |      |   2301612206 | 19/B20021F0 | 19/B2002228         | reserved   |              
(1 row)

### What did you expect to happen?

- logical replication syncronization slots must not be dropped
- unknown permanent slots must be dropped still it is not clear at which moment

### Patroni/PostgreSQL/DCS version

-   Patroni version: 2.1.1, 3.3.0
-   PostgreSQL version: 15 and higher
-   DCS (and its version): etcd v3.3.11

### Patroni configuration file

```yaml
$ cat postgres.yml 
scope: clustername
name: source_host

restapi:
...

etcd:
...

bootstrap:
    dcs:
        ttl: 135
        retry_timeout: 60
        loop_wait: 10
        maximum_lag_on_failover: 1048576
        synchronous_mode: true
        synchronous_mode_strict: false
        postgresql:
            parameters:
                max_connections: '110'
                max_locks_per_transaction: '64'
                max_worker_processes: '32'
                max_prepared_transactions: '12'
                wal_level: logical
                wal_log_hints: on
                track_commit_timestamp: off
                wal_keep_size: 256MB
                max_wal_senders: '10'
                max_replication_slots: '10'
                logical_decoding_work_mem: 64MB
                auth_activity_period: '60'
                authentication_max_workers: '16'
            use_pg_rewind: true
            use_slots: true
    initdb:
      - encoding: UTF8
      - data-checksums
postgresql:
...
    parameters:

#defaults

tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
```

### patronictl show-config

```yaml
$ patronictl -c /etc/patroni/postgres.yml show-config
loop_wait: 10
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    auth_activity_period: '60'
    authentication_max_workers: '16'
    logical_decoding_work_mem: 64MB
    max_connections: '110'
    max_locks_per_transaction: '64'
    max_prepared_transactions: '12'
    max_replication_slots: '2'
    max_wal_senders: '10'
    max_worker_processes: '32'
    track_commit_timestamp: false
    wal_keep_size: 64MB
    wal_level: logical
    wal_log_hints: true
  use_pg_rewind: true
  use_slots: true
retry_timeout: 60
synchronous_mode: true
synchronous_mode_strict: false
ttl: 135
```

### Patroni log files

```shell
-- patroni log: it does not inform that slot pg_19732_sync_17854_7442686389907653644 was deleted
Dec  4 10:50:52 source_host patroni.bin[1309802]: 2024-12-04 10:50:52,755 ERROR: Failed to drop replication slot 'test_sub1'
Dec  4 10:50:52 source_host patroni.bin[1309802]: 2024-12-04 10:50:52,819 INFO: no action. I am (srv-0-213.db.dev.sbt) the leader with the lock
```

### PostgreSQL log files

```shell
-- postgresql log: create slot by user and delete slot by patroni
2024-12-04 10:48:31 MSK [2249777]: [9-1] app=psql,user=postgres,db=postgres,client=source_host,type=client backend LOG:  duration: 10.285 ms  statement: SELECT * FROM pg_create_logical_replication_slot('pg_19732_sync_17854_7442686389907653644', 'pgoutput', 'false');
2024-12-04 10:48:32 MSK [2249641]: [57-1] app=Patroni,user=patroni,db=postgres,client=127.0.0.1,type=client backend LOG:  duration: 0.523 ms  statement: SELECT CASE WHEN pg_catalog.pg_is_in_recovery() THEN 0 ELSE ('x' || pg_catalog.substr(pg_catalog.pg_walfile_name(pg_catalog.pg_current_wal_lsn()), 1, 8))::bit(32)::int END, CASE WHEN pg_catalog.pg_is_in_recovery() THEN 0 ELSE pg_catalog.pg_wal_lsn_diff(pg_catalog.pg_current_wal_lsn(), '0/0')::bigint END, pg_catalog.pg_wal_lsn_diff(pg_catalog.pg_last_wal_replay_lsn(), '0/0')::bigint, pg_catalog.pg_wal_lsn_diff(COALESCE(pg_catalog.pg_last_wal_receive_lsn(), '0/0'), '0/0')::bigint, pg_catalog.pg_is_in_recovery() AND pg_catalog.pg_is_wal_replay_paused(), 0, CASE WHEN latest_end_lsn IS NULL THEN NULL ELSE received_tli END, slot_name, conninfo, NULL FROM pg_catalog.pg_stat_get_wal_receiver()
2024-12-04 10:48:32 MSK [2249641]: [58-1] app=Patroni,user=patroni,db=postgres,client=127.0.0.1,type=client backend LOG:  duration: 0.538 ms  statement: SELECT slot_name, slot_type, plugin, database, datoid, catalog_xmin, pg_catalog.pg_wal_lsn_diff(confirmed_flush_lsn, '0/0')::bigint FROM pg_catalog.pg_replication_slots
2024-12-04 10:48:32 MSK [2249641]: [59-1] app=Patroni,user=patroni,db=postgres,client=127.0.0.1,type=client backend LOG:  duration: 1.551 ms  statement: SELECT pg_catalog.pg_drop_replication_slot('pg_19732_sync_17854_7442686389907653644') WHERE EXISTS (SELECT 1 FROM pg_catalog.pg_replication_slots WHERE slot_name = 'pg_19732_sync_17854_7442686389907653644' AND NOT active)
```

### Have you tried to use GitHub issue search?

- [X] Yes

**Comment by hughcapet** [comment]
> From one side any unknown slot (but temporary one or ignore_slots) is about to be dropped, from another side it is often not dropped until restart. It is not clear at which moment a slot is expected to be dropped which is confusing.
 
Slot that is not matching the ones in the `slots` section is dropped once it is detected by Patroni and is not active (you can not drop active slots by PG design). Pause mode prevents slots deletion. Your reproduction is the perfect illustration of that.
What do you mean by "often not dropped until restart."?

**Comment by akalgush** [comment]
Thanks for your answer.
We figured out that any inactive unknow slot once created is not dropped until restart. The test shows a special case where exist an unknown active slot which is not touched until pause/resume. Then patroni tries to drop it every loop_wait interval and not only this one... All the rest inactive slots are dropped once created on the next loop_wait cycle. This causes a problem for logical replication slots specifically the ones created for initial table syncronization. The table syncronization is made in two phases: COPY (which could be quite time-consuming) and then replicating DML changes made during copying. For the second phase a subscription just creates a slot with a special name and start the replication on it once the copy is finished. However patroni already deleted this slot while the copy was in process. 
From docs: https://www.postgresql.org/docs/current/logical-replication-subscription.html
.
'Additional table synchronization slots are normally transient, created internally to perform initial table synchronization and dropped automatically when they are no longer needed. These table synchronization slots have generated names: “pg_%u_sync_%u_%llu” (parameters: Subscription oid, Table relid, system identifier sysid)'
.
These slots used to be temporary before postgresql v14 so were never touched by patroni. Nowadays they are in trouble in special cases like the one provided above. Logical replication just can not complete an initial table syncronization in this case.

**Comment by hughcapet** [comment]
i described how it works. you also mentioned you know about the "ignore_slots" configuration used to prevent deletion of the manually created slots and "pause mode" that disables the slots management. So what is the issue here?

> We figured out that any inactive unknow slot once created is not dropped until restart. The test shows a special case where exist an unknown active slot which is not touched until pause/resume.

I still do not get which slot form your example you mean.
test_sub1 is active -- Patroni fails to delete it
pg_19732_sync_17854_7442686389907653644 -- inactive, slot is deleted immediately

Is the problem that you could not find log line about pg_19732_sync_17854_7442686389907653644 being deleted?

**Comment by glushakov** [comment]
We also encountered this problem
1. create a permanent slot in patroni for logical replication
2. create a subscription specifying the slot name
After that additional synchronization slots are created on the publisher, which patroni does not know about (and the administrator does not know name of slots and cannot create them in patroni in advance)

If patroni is restarted, it will delete the synchronization slot and in most cases the subscription will have to be recreated.

As an workaround, maybe teach patroni to ignore slots by the mask *pg_*_sync* ?

**Comment by akalgush** [comment]
Exactly!
Slot pg_19732_sync_17854_7442686389907653644 must not be deleted by Patroni until restart I think. The name of this slot is unpredictable so can not be included in the list of the ones which are ignored.
Slot test_sub1 is the one to make patroni start deleting it after resume. Even if it is unknown to patroni it is not prohibited to create it like this. If patroni does not in the cycle of deleting any slot like test_sub1 then it does not touch pg_*_sync_* one until restart.
This issue is about a proper patroni support for logical replication since v14 and above.
Again the doc on the following page is not completely obvious about the way how slots are being managed and cause a huge amount of gossips: https://patroni.readthedocs.io/en/latest/dynamic_configuration.html

**Comment by akalgush** [comment]
Just wondering if the following articles might make a sense...
https://amitkapila16.blogspot.com/2021/09/logical-replication-improvements-in.html
https://www.postgresql.fastware.com/blog/logical-replication-tablesync-workers
https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=ce0fdbfe9722867b7fad4d3ede9b6a6bfc51fb4e

**Comment by hughcapet** [comment]
Ok. Now I get it. I missed the use case and that “generated” part in your message while concentrating on your example:)

You can try this in ignore _slots to prevent deletion of any slots in this db:

```
  - type: logical
    database: my_db
    plugin: pgoutput
```


And yes, I agree, documentation can be definitely improved. Or maybe have a separate section on slots management

**Comment by akalgush** [comment]
Thank you for this workaround! It definitely works. However, it does not sound a very great solution on constant basis as it might cause 'not-deleting' any wrong slots on restart/switchover, I guess.. The filter it provides is too wide probably.. So, it is great to have it still wondering if there could be any constant decisions..

**Comment by hughcapet** [comment]
Yes, i agree. Ability to define mask sounds right to me. But maybe @CyberDem0n has another opinion
